### PR TITLE
Allow incremental servicing of packages and more packaging infra cleanup

### PR DESCRIPTION
**Pull request by ViktorHofer**
Requires https://github.com/dotnet/arcade/pull/6728
Based on the discussion in https://github.com/dotnet/runtime/pull/46180

This PR further cleans up the packaging infra and allows "incremental servicing" at project level instead of traversal level (as it was before for libraries). By default, nuget csprojs or pkgprojs will be built if the `<GeneratePackage />` property is set to true.

The `GeneratePackage` property is currently conditioned as:
```
    <GeneratePackage Condition="(('$(PreReleaseVersionLabel)' != 'servicing' and
                                  '$(RepositoryName)' != 'runtimelab') or
                                 '$(BuildAllProjects)' != 'true')">true</GeneratePackage>
```
This means that during servicing or as part of runtimelab, root builds won't generate packages unless a project opts in via `<GeneratePackage />`. `<BuildAllProjects />` identifies a builds as a root build (vs an individual project build: `dotnet pack ...`).

I wasn't sure how to condition runtimelab so I used the `RepositoryName` property for now.

### CODE REVIEW

File: eng/packaging.targets [hunk 1]
```diff
@@ -27,12 +27,13 @@
...
-          Condition="'$(IsSourceProject)' == 'true' and
-                     '$(IsPackable)' == 'true' and
+  <!-- Point to the custom pack target which invokes the pkgproj from a source csproj. -->
+  <Target Name="_OverridePackDependsOnForCsProjToPkgProj"
+          Condition="'$(GeneratePackage)' == 'true' and
+                     '$(IsSourceProject)' == 'true' and
                      Exists('$(PkgProjPath)')">
     <PropertyGroup>
-      <!-- Point to the custom pack target which invokes the pkgproj. -->
+      <IsPackable>true</IsPackable>
```
Comment (Anipik) [inline eng/packaging.targets:36]: why are we setting IsPackable to true here ?

**Review by Anipik — APPROVED** [review]
Packing changes for the libraries looks good. I believe there are no functional change in coreclr related changes. its seem like just refactoring ?

**Comment by ViktorHofer** [comment]
The change applies to all packages in the repo that use the nuget pack task or pkgprojs. It means that no packages will be generated in servicing by default unless we set the <Generate Package /> to true in their project file. I'm unsure which packages we always want to generate but I assume the sfx ones?

**Comment by ViktorHofer** [comment]
Timeout is https://github.com/dotnet/runtime/issues/46484

File: Directory.Build.props [hunk 2]
```diff
@@ -220,14 +220,19 @@
...
+    <SuppressFinalPackageVersion Condition="'$(SuppressFinalPackageVersion)' == '' and ($(IsExperimentalAssembly) or $(IsPrivateAssembly))">true</SuppressFinalPackageVersion>
     <IsShippingAssembly Condition="$(IsExperimentalAssembly)">false</IsShippingAssembly>
     <!-- We don't want Private packages to be shipped to NuGet.org -->
     <IsShippingPackage Condition="($(MSBuildProjectName.Contains('Private')) or $(IsExperimentalAssembly)) and '$(MSBuildProjectExtension)' == '.pkgproj'">false</IsShippingPackage>
-    <!-- Set IsPackable true for all pkgproj as Arcade unconditionally overrides the value. -->
-    <IsPackable Condition="'$(MSBuildProjectExtension)' == '.pkgproj'">true</IsPackable>
+    <!-- A package isn't generated (in traversal builds) if in servicing or in runtimelab. Intended to be overridden at project level. -->
+    <GeneratePackage Condition="'$(GeneratePackage)' == ''">true</GeneratePackage>
+    <GeneratePackage Condition="(('$(PreReleaseVersionLabel)' == 'servicing' or
+                                  '$(RepositoryName)' == 'runtimelab') and
```
Comment (jkotas) [inline Directory.Build.props:232]: How is the `RepositoryName` meant to be set in runtimelab?

**Pull request merged**